### DISCUSSION TITLE
Packaging CLI application release for Homebrew & GitHub

### DISCUSSION THREAD

**Discussion by ahmedsaheed** [jreleaser/jreleaser#851] [Q&A] [answered]
Hi, 

I have recently created a CLI application with picocli and everything works fine on my local machcine. 
I am now struggling with packaging my apllication on homebrew using Jreleaser.

I am a bit overwhelmed with the steps involved on those documentations and would need some step by step clerification on how to 
get this accomplished.

What i have accomplished

- I have successfully added Jrealeser to my project
- Also i have set up the release for both GitHub and Homebrew [see here](https://github.com/ahmedsaheed/quranCLI/blob/main/build.gradle)
- The realease works for GitHub [here](https://github.com/ahmedsaheed/quranCLI/releases) but doesn't work for homebrew

What i need to know 

- Would i have to do any GraalVM configuration ? And if yes, what're the steps
- How can i correctly cofigure my package for homebrew.

Thanks a ton,
Ahmed.

**Comment by aalmiray** [comment]
Hi Ahmed, thank you for trying out JReleaser. I see that in your configuration you set up a lot of settings, most of them use default values evaluated automatically by the model. You can remove those settings to reduce verbosity. 

Now on to brew configuration. You have both brew and cask settings. If your application is just a zip then I suggest to remove the cask configuration, as that is mostly used for formulas that rely on installers (.dmg and .pkg). Also, the value for `token` is explicitly set to "_DO_NOT_SET_HERE_" which is an invalid value. You must define and environment variable such as JRELEASER_HOMEBREW_GITHUB_TOKEN or JRELEASER_GITHUB_TOKEN but make sure that the token in use has write access to other repositories, as explained at https://jreleaser.org/guide/latest/configuration/packagers/homebrew.html#_tokens

**Comment by ahmedsaheed** [comment]
Thank you very much for your  response!
I have removed the previously set token value and it  currently uses env variable set in config.properties.

Also I have gotten rid of cask configuration as it's of no use.
But when i run ` ./gradlew jreleaserFullRelease` i get **Build Failed with an unexpected exception**. Although `./gradlew jreleaserConfig` is always sucessful. What could be the issue as the error is a little bit vague.

Also, my formular name does change even after specifying `formulaName = 'quranCLI'` it still shows as `App`
> See Full Message here

```
> Task :jreleaserFullRelease
[INFO]  JReleaser 1.0.0
[INFO]    - basedir set to /Users/ahmedsaheed/Desktop/Desktop/quranCLI
[INFO]  Loading variables from /Users/ahmedsaheed/.jreleaser/config.properties
[INFO]  Validating configuration
[INFO]  Project version set to 1.0
[INFO]  Release is not snapshot
[INFO]  Timestamp is 2022-06-29T00:28:18.954623+01:00
[INFO]  HEAD is at ed9895b
[INFO]  Platform is osx-x86_64
[INFO]  dry-run set to false
[WARN]  Cannot parse version 'v1.0.0'
[WARN]  Cannot parse version 'early-access'
[INFO]  Generating changelog: build/jreleaser/release/CHANGELOG.md
[INFO]  Calculating checksums
[INFO]    [checksum] All checksums are up-to-date. Skipping
[INFO]  Signing files
[INFO]    [sign] Signing is not enabled. Skipping
[INFO]  Uploading is not enabled. Skipping
[INFO]  Releasing to https://github.com/ahmedsaheed/quranCLI
[INFO]   - uploading quranCLI-1.0.zip
[INFO]   - uploading checksums_sha256.txt
[WARN]  Could not update release v1.0.0 with discussion category Announcements
[INFO]  Preparing distributions
[INFO]    - Preparing app distribution
[INFO]      [brew] preparing app distribution
[INFO]  Packaging distributions
[INFO]    - Packaging app distribution
[INFO]      [brew] packaging app distribution
[INFO]  Publishing distributions
[INFO]    - Publishing app distribution
[INFO]      [brew] publishing app distribution
[INFO]      [brew] setting up repository ahmedsaheed/ahmedsaheed
[INFO]  Writing output properties to build/jreleaser/output.properties
[ERROR] JReleaser failed after 8.623 s

> Task :jreleaserFullRelease FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jreleaserFullRelease'.
> Unexpected error

```

I have also updated the build.gradle file so you could take a look from the previous url.

Please pardon my disturbance, thanks

**Comment by aalmiray** [comment, marked as answer]
Hi Ahmed,

Could you try invoking `gradle jreleaserFullRelease -S`? This will print out the full stack trace. Also, `trace.log` should have the stack trace as well. I had a look at your project and the minimum configuration for it (using default values) would be:

```groovy
jreleaser {
    release {
        github {
            overwrite = true
        }
    }

    distributions {
        qurancli {
            active = 'ALWAYS'
            name = 'qurancli'
            artifact {
                path = 'build/distributions/{{distributionName}}-{{projectVersion}}.zip'
            }
        }
    }

    packagers {
        brew {
            active = 'ALWAYS'
            commitAuthor {
                name = 'Ahmed Saheed'
                email = 'saheed@quranCLI.com'
            }
        }
    }
}
```

Make sure to set the project version to `1.0.0` as it currently is set to `1.0`. With these settings I've got the following

```
$ $ tree build/jreleaser/
build/jreleaser/
├── checksums
│   ├── checksums_sha256.txt
│   └── qurancli
│       └── qurancli-1.0.0.zip.sha256
├── output.properties
├── prepare
│   └── qurancli
│       └── brew
│           ├── Formula
│           │   └── qurancli.rb
│           └── README.md
└── trace.log
```

```
$ cat build/jreleaser/prepare/qurancli/brew/Formula/qurancli.rb 
# Generated with JReleaser 1.0.0 at 2022-06-29T11:26:49.692526+02:00
class Qurancli < Formula
  desc "A simple tool to Read, Search and Recite the Quran."
  homepage "https://github.com/ahmedsaheed/quranCLI"
  url "https://github.com/ahmedsaheed/quranCLI/releases/download/v1.0.0/qurancli-1.0.0.zip"
  version "1.0.0"
  sha256 "bdcb9abc160d96c438f294ad87be09e89c864f83966e41d5ebc380b79f4dcd94"
  license "Apache-2.0"

  depends_on "openjdk@17"

  def install
    libexec.install Dir["*"]
    bin.install_symlink "#{libexec}/bin/qurancli" => "qurancli"
  end

  test do
    output = shell_output("#{bin}/qurancli --version")
    assert_match "1.0.0", output
  end
end
```

Another thing I noticed is that the main class is found in the default package, it's recommended that you define your own package, such as `org.saheed.ahmed`, or `org.qurancli`, or another value based on a reverse domain under your control. It could even be `com.github.ahmedsaheed`.

**Reply by ahmedsaheed** [reply]
Thanks very much for your support, I was able to get it working after changing the `formularName = quranCLI` to `formulaName = qurancli`

Then the `qurancli.rb ` was created in a repo named`ahmedsaheed/ahmedsaheed` which i manually  changed to `ahmedsaheed/homebrew-tap`

Thanks once again.
